src/webots/vrml/WbNode.cpp

```diff
@@ -1881,6 +1903,264 @@ bool WbNode::isProtoParameterNode() const {
+  for (int i = 0; i < internalProtoNodes.size(); ++i) {
+    WbNode *n = internalProtoNodes[i]->protoParameterNode();
+
+    while (n != NULL) {
+      bool added = false;
+      for (int j = 0; j < invisibleProtoParameterNodes.size(); ++j)
+        if (n->level() > invisibleProtoParameterNodes[j]->level()) {  // insert them from lowest to highest level
+          invisibleProtoParameterNodes.insert(j, n);
+          added = true;
+          break;  // need to break otherwise invisibleProtoParameterNodes grows infinitly
+        }
+
+      if (!added)
+        invisibleProtoParameterNodes.append(n);
+
+      n = n->protoParameterNode();
+    }
+  }
+
+  /*
+  // TODO: remove before merge
+  printf("\nINVISIBLE PROTO PARAMETER NODES (WHAT WILL BE REMOVED)\n");
+  for (int i = 0; i < invisibleProtoParameterNodes.size(); ++i) {
+    printf("  [L%d] %s [%p]\n", invisibleProtoParameterNodes[i]->level(),
+           invisibleProtoParameterNodes[i]->usefulName().toUtf8().constData(), invisibleProtoParameterNodes[i]);
+  }
+  */
+
+  if (invisibleProtoParameterNodes.size() == 0)
+    return;
+
+  // break link between [field] -> [parameter] and [internal node] -> [parameter node] (from internal node side)
+  for (int i = 0; i < internalProtoNodes.size(); ++i) {
+    internalProtoNodes[i]->disconnectInternalNode();
+    const QVector<WbField *> fields = internalProtoNodes[i]->fields();
+
+    for (int j = 0; j < fields.size(); j++)
+      fields[j]->setParameter(NULL);
+
+    internalProtoNodes[i]->setProtoParameterNode(NULL);  // break link with proto parameter node
+  }
+
+  // break link [parameter] -> [internal field] and [parameter node] -> [internal node] (from parameter node side)
+  for (int i = 0; i < invisibleProtoParameterNodes.size(); ++i) {
+    invisibleProtoParameterNodes[i]->clearProtoParameterNodeInstances();  // clear downward references
+
+    // clear internal field references (for protoParameterNodes the reference is kept in its fields)
+    QVector<WbField *> fields = invisibleProtoParameterNodes[i]->fields();
+    for (int i = 0; i < fields.size(); ++i)
+      fields[i]->clearInternalFields();
+  }
+
+  // now the proto parameter nodes can be deleted, depending on the situation it can either be in the parameter or field side of
+  // the parent node. The signal is not emitted to prevent the internal node from being deleted as well in the process
+  for (int i = 0; i < invisibleProtoParameterNodes.size(); ++i) {
+    WbNode *parameterNode = invisibleProtoParameterNodes[i];
+    WbNode *parent = parameterNode->parentNode();
+
+    const QVector<WbField *> fields = parent->fields();
+    for (int j = 0; j < fields.size(); ++j) {
+      WbSFNode *sfnode = dynamic_cast<WbSFNode *>(fields[j]->value());
+      WbMFNode *mfnode = dynamic_cast<WbMFNode *>(fields[j]->value());
+      if (sfnode && sfnode->value() == parameterNode) {
+        sfnode->setValueNoSignal(NULL);
```

Comment:
Note that it is not necessary to duplicate the method so that it doesn't emit the signal, but you can simply use:
```
sfnode->blockSignals(true);
sfnode->setValue(NULL);
sfnode->blockSignals(false);
```

The methods that don't emit signals are needed when called indirectly by slots where it is not possible to block the signals.